File: frontend/src/lib/routePlanValidation.ts
import type { RouteResult } from "@/lib/types";

export type ValidationBand = "green" | "yellow" | "red";

export interface RouteValidationThresholds {
  max_buy_drift_pct: number;
  max_sell_drift_pct: number;
  min_route_profit_retained_pct: number;
  min_stop_liquidity_retained_pct: number;
}

export interface RoutePlanStopValidation {
  stop_key: string;
  buy_ceiling_isk: number;
  sell_floor_isk: number;
  buy_drift_pct: number;
  sell_drift_pct: number;
  retained_profit_pct: number;
  liquidity_retained_pct: number;
  band: ValidationBand;
}

export interface RoutePlanValidationResult {
  band: ValidationBand;
  snapshot_stale: boolean;
  total_buy_drift_pct: number;
  total_sell_drift_pct: number;
  route_profit_retained_pct: number;
  min_stop_liquidity_retained_pct: number;
  checkpoints: Array<{ name: "pre-undock" | "pre-sale"; band: ValidationBand }>;
  stops: RoutePlanStopValidation[];
}

const bandOrder: Record<ValidationBand, number> = {
  green: 0,
  yellow: 1,
  red: 2,
};

function maxBand(a: ValidationBand, b: ValidationBand): ValidationBand {
  return bandOrder[a] >= bandOrder[b] ? a : b;
}

function bandByMax(actual: number, threshold: number): ValidationBand {
  if (threshold <= 0 || actual <= threshold) return "green";
  if (actual <= threshold * 1.2) return "yellow";
  return "red";
}

function bandByMin(actual: number, threshold: number): ValidationBand {
  if (threshold <= 0 || actual >= threshold) return "green";
  if (actual >= threshold * 0.8) return "yellow";
  return "red";
}

function pctIncrease(base: number, current: number): number {
  if (base <= 0) return current > 0 ? 100 : 0;
  return ((current - base) / base) * 100;
}

function pctDown(base: number, current: number): number {
  if (base <= 0 || current >= base) return 0;
  return ((base - current) / base) * 100;
}

function retained(base: number, current: number): number {
  if (base <= 0) return 100;
  return (current / base) * 100;
}

export function evaluateRoutePlanValidation(input: {
  route: RouteResult;
  thresholds: RouteValidationThresholds;
  nowMs?: number;
  staleAfterMs?: number;
}): RoutePlanValidationResult {
  const nowMs = input.nowMs ?? Date.now();
  const staleAfterMs = input.staleAfterMs ?? 15 * 60 * 1000;

  const stops = input.route.Hops.map((hop, idx) => {
    const units = Math.max(0, Math.trunc(hop.Units ?? 0));
    const modeled = Math.max(1, Math.trunc((hop.modeled_qty ?? units) || 1));
    const snapBuy = units * (hop.BuyPrice ?? 0);
    const snapSell = units * (hop.SellPrice ?? 0);
    const currentBuy = units * (hop.effective_buy ?? hop.BuyPrice ?? 0);
    const currentSell = units * (hop.effective_sell ?? hop.SellPrice ?? 0);

    const buyLiquidityRetained = retained(
      modeled,
      Math.max(0, hop.buy_remaining ?? modeled),
    );
    const sellLiquidityRetained = retained(
      modeled,
      Math.max(0, hop.sell_remaining ?? modeled),
    );
    const liquidityRetained = Math.min(
      buyLiquidityRetained,
      sellLiquidityRetained,
    );

    const buyDrift = pctIncrease(snapBuy, currentBuy);
    const sellDrift = pctDown(snapSell, currentSell);
    const retainedProfit = retained(
      snapSell - snapBuy,
      currentSell - currentBuy,
    );

    let band: ValidationBand = "green";
    band = maxBand(
      band,
      bandByMax(buyDrift, input.thresholds.max_buy_drift_pct),
    );
    band = maxBand(
      band,
      bandByMax(sellDrift, input.thresholds.max_sell_drift_pct),
    );
    band = maxBand(
      band,
      bandByMin(retainedProfit, input.thresholds.min_route_profit_retained_pct),
    );
    band = maxBand(
      band,
      bandByMin(
        liquidityRetained,
        input.thresholds.min_stop_liquidity_retained_pct,
      ),
    );

    return {
      stop_key: `${hop.SystemID}:${hop.DestSystemID}:${idx}`,
      buy_ceiling_isk: snapBuy * (1 + input.thresholds.max_buy_drift_pct / 100),
      sell_floor_isk:
        snapSell * (1 - input.thresholds.max_sell_drift_pct / 100),
      buy_drift_pct: buyDrift,
      sell_drift_pct: sellDrift,
      retained_profit_pct: retainedProfit,
      liquidity_retained_pct: liquidityRetained,
      band,
      snapBuy,
      snapSell,
      currentBuy,
      currentSell,
    };
  });

  const totals = stops.reduce(
    (acc, stop) => {
      acc.snapBuy += stop.snapBuy;
      acc.snapSell += stop.snapSell;
      acc.currentBuy += stop.currentBuy;
      acc.currentSell += stop.currentSell;
      acc.minLiquidity = Math.min(
        acc.minLiquidity,
        stop.liquidity_retained_pct,
      );
      return acc;
    },
    {
      snapBuy: 0,
      snapSell: 0,
      currentBuy: 0,
      currentSell: 0,
      minLiquidity: 100,
    },
  );

  const totalBuyDrift = pctIncrease(totals.snapBuy, totals.currentBuy);
  const totalSellDrift = pctDown(totals.snapSell, totals.currentSell);
  const routeProfitRetained = retained(
    totals.snapSell - totals.snapBuy,
    totals.currentSell - totals.currentBuy,
  );

  const snapshotTs = input.route.Hops.map((hop) =>
    hop.snapshot_ts ? Date.parse(hop.snapshot_ts) : Number.NaN,
  )
    .filter((ts) => Number.isFinite(ts))
    .reduce((min, ts) => Math.min(min, ts), Number.POSITIVE_INFINITY);
  const snapshotStale = Number.isFinite(snapshotTs)
    ? nowMs - snapshotTs > staleAfterMs
    : false;

  let preUndock = maxBand(
    bandByMax(totalBuyDrift, input.thresholds.max_buy_drift_pct),
    bandByMin(
      totals.minLiquidity,
      input.thresholds.min_stop_liquidity_retained_pct,
    ),
  );
  let preSale = maxBand(
    bandByMax(totalSellDrift, input.thresholds.max_sell_drift_pct),
    bandByMin(
      routeProfitRetained,
      input.thresholds.min_route_profit_retained_pct,
    ),
  );
  if (snapshotStale) {
    preUndock = maxBand(preUndock, "yellow");
    preSale = maxBand(preSale, "yellow");
  }

  return {
    band: maxBand(preUndock, preSale),
    snapshot_stale: snapshotStale,
    total_buy_drift_pct: totalBuyDrift,
    total_sell_drift_pct: totalSellDrift,
    route_profit_retained_pct: routeProfitRetained,
    min_stop_liquidity_retained_pct: totals.minLiquidity,
    checkpoints: [
      { name: "pre-undock", band: preUndock },
      { name: "pre-sale", band: preSale },
    ],
    stops: stops.map(
      ({
        snapBuy: _sb,
        snapSell: _ss,
        currentBuy: _cb,
        currentSell: _cs,
        ...rest
      }) => rest,
    ),
  };
}

File: internal/engine/route_plan_validation.go
package engine

import "time"

type ValidationBand string

const (
	ValidationBandPass ValidationBand = "pass"
	ValidationBandWarn ValidationBand = "warn"
	ValidationBandFail ValidationBand = "fail"
)

type RoutePlanValidationThresholds struct {
	MaxBuyDriftPct              float64 `json:"max_buy_drift_pct"`
	MaxSellDriftPct             float64 `json:"max_sell_drift_pct"`
	MinRouteProfitRetainedPct   float64 `json:"min_route_profit_retained_pct"`
	MinStopLiquidityRetainedPct float64 `json:"min_stop_liquidity_retained_pct"`
}

type RoutePlanValidationStop struct {
	StopKey               string  `json:"stop_key"`
	SnapshotBuyTotalISK   float64 `json:"snapshot_buy_total_isk"`
	SnapshotSellTotalISK  float64 `json:"snapshot_sell_total_isk"`
	SnapshotBuyLiquidity  float64 `json:"snapshot_buy_liquidity"`
	SnapshotSellLiquidity float64 `json:"snapshot_sell_liquidity"`
	CurrentBuyTotalISK    float64 `json:"current_buy_total_isk"`
	CurrentSellTotalISK   float64 `json:"current_sell_total_isk"`
	CurrentBuyLiquidity   float64 `json:"current_buy_liquidity"`
	CurrentSellLiquidity  float64 `json:"current_sell_liquidity"`
}

type RoutePlanValidationRequest struct {
	SnapshotAt  time.Time                     `json:"snapshot_at"`
	ValidatedAt time.Time                     `json:"validated_at"`
	StaleAfter  time.Duration                 `json:"stale_after"`
	Thresholds  RoutePlanValidationThresholds `json:"thresholds"`
	Stops       []RoutePlanValidationStop     `json:"stops"`
}

type RoutePlanValidationStopResult struct {
	StopKey              string         `json:"stop_key"`
	BuyCeilingISK        float64        `json:"buy_ceiling_isk"`
	SellFloorISK         float64        `json:"sell_floor_isk"`
	BuyDriftPct          float64        `json:"buy_drift_pct"`
	SellDriftPct         float64        `json:"sell_drift_pct"`
	LiquidityRetainedPct float64        `json:"liquidity_retained_pct"`
	Band                 ValidationBand `json:"band"`
}

type RoutePlanValidationCheckpoint struct {
	Name     string         `json:"name"`
	Band     ValidationBand `json:"band"`
	Messages []string       `json:"messages,omitempty"`
}

type RoutePlanValidationResponse struct {
	Band                     ValidationBand                  `json:"band"`
	SnapshotStale            bool                            `json:"snapshot_stale"`
	TotalBuyDriftPct         float64                         `json:"total_buy_drift_pct"`
	TotalSellDriftPct        float64                         `json:"total_sell_drift_pct"`
	RouteProfitRetainedPct   float64                         `json:"route_profit_retained_pct"`
	MinStopLiquidityRetained float64                         `json:"min_stop_liquidity_retained"`
	Checkpoints              []RoutePlanValidationCheckpoint `json:"checkpoints"`
	Stops                    []RoutePlanValidationStopResult `json:"stops"`
}

func ValidateRoutePlan(req RoutePlanValidationRequest) RoutePlanValidationResponse {
	staleAfter := req.StaleAfter
	if staleAfter <= 0 {
		staleAfter = 15 * time.Minute
	}
	snapshotStale := !req.SnapshotAt.IsZero() && req.ValidatedAt.Sub(req.SnapshotAt) > staleAfter

	totalSnapBuy, totalSnapSell := 0.0, 0.0
	totalCurBuy, totalCurSell := 0.0, 0.0
	minLiquidityRetained := 100.0
	stopResults := make([]RoutePlanValidationStopResult, 0, len(req.Stops))
	for _, stop := range req.Stops {
		totalSnapBuy += stop.SnapshotBuyTotalISK
		totalSnapSell += stop.SnapshotSellTotalISK
		totalCurBuy += stop.CurrentBuyTotalISK
		totalCurSell += stop.CurrentSellTotalISK

		buyDrift := pctDeltaUp(stop.SnapshotBuyTotalISK, stop.CurrentBuyTotalISK)
		sellDrift := pctDeltaDown(stop.SnapshotSellTotalISK, stop.CurrentSellTotalISK)
		liquidityRetained := minFloat(
			retainedPct(stop.SnapshotBuyLiquidity, stop.CurrentBuyLiquidity),
			retainedPct(stop.SnapshotSellLiquidity, stop.CurrentSellLiquidity),
		)
		if liquidityRetained < minLiquidityRetained {
			minLiquidityRetained = liquidityRetained
		}

		band := ValidationBandPass
		band = maxBand(band, bandByMax(buyDrift, req.Thresholds.MaxBuyDriftPct))
		band = maxBand(band, bandByMax(sellDrift, req.Thresholds.MaxSellDriftPct))
		band = maxBand(band, bandByMin(liquidityRetained, req.Thresholds.MinStopLiquidityRetainedPct))

		stopResults = append(stopResults, RoutePlanValidationStopResult{
			StopKey:              stop.StopKey,
			BuyCeilingISK:        stop.SnapshotBuyTotalISK * (1 + req.Thresholds.MaxBuyDriftPct/100),
			SellFloorISK:         stop.SnapshotSellTotalISK * (1 - req.Thresholds.MaxSellDriftPct/100),
			BuyDriftPct:          buyDrift,
			SellDriftPct:         sellDrift,
			LiquidityRetainedPct: liquidityRetained,
			Band:                 band,
		})
	}
	if len(req.Stops) == 0 {
		minLiquidityRetained = 100
	}

	totalBuyDrift := pctDeltaUp(totalSnapBuy, totalCurBuy)
	totalSellDrift := pctDeltaDown(totalSnapSell, totalCurSell)
	snapshotProfit := totalSnapSell - totalSnapBuy
	currentProfit := totalCurSell - totalCurBuy
	retainedProfit := retainedPct(snapshotProfit, currentProfit)

	preUndockBand := maxBand(
		bandByMax(totalBuyDrift, req.Thresholds.MaxBuyDriftPct),
		bandByMin(minLiquidityRetained, req.Thresholds.MinStopLiquidityRetainedPct),
	)
	preSaleBand := maxBand(
		bandByMax(totalSellDrift, req.Thresholds.MaxSellDriftPct),
		bandByMin(retainedProfit, req.Thresholds.MinRouteProfitRetainedPct),
	)
	if snapshotStale {
		preUndockBand = maxBand(preUndockBand, ValidationBandWarn)
		preSaleBand = maxBand(preSaleBand, ValidationBandWarn)
	}

	response := RoutePlanValidationResponse{
		Band:                     maxBand(preUndockBand, preSaleBand),
		SnapshotStale:            snapshotStale,
		TotalBuyDriftPct:         totalBuyDrift,
		TotalSellDriftPct:        totalSellDrift,
		RouteProfitRetainedPct:   retainedProfit,
		MinStopLiquidityRetained: minLiquidityRetained,
		Checkpoints: []RoutePlanValidationCheckpoint{
			{Name: "pre-undock", Band: preUndockBand, Messages: checkpointMessagesPreUndock(snapshotStale, totalBuyDrift, minLiquidityRetained, req.Thresholds)},
			{Name: "pre-sale", Band: preSaleBand, Messages: checkpointMessagesPreSale(snapshotStale, totalSellDrift, retainedProfit, req.Thresholds)},
		},
		Stops: stopResults,
	}
	return response
}

func checkpointMessagesPreUndock(stale bool, buyDrift, minLiquidity float64, th RoutePlanValidationThresholds) []string {
	messages := []string{}
	if stale {
		messages = append(messages, "snapshot_stale")
	}
	if band := bandByMax(buyDrift, th.MaxBuyDriftPct); band != ValidationBandPass {
		messages = append(messages, "buy_drift")
	}
	if band := bandByMin(minLiquidity, th.MinStopLiquidityRetainedPct); band != ValidationBandPass {
		messages = append(messages, "liquidity_retention")
	}
	return messages
}

func checkpointMessagesPreSale(stale bool, sellDrift, retained float64, th RoutePlanValidationThresholds) []string {
	messages := []string{}
	if stale {
		messages = append(messages, "snapshot_stale")
	}
	if band := bandByMax(sellDrift, th.MaxSellDriftPct); band != ValidationBandPass {
		messages = append(messages, "sell_drift")
	}
	if band := bandByMin(retained, th.MinRouteProfitRetainedPct); band != ValidationBandPass {
		messages = append(messages, "profit_retained")
	}
	return messages
}

func pctDeltaUp(base, current float64) float64 {
	if base <= 0 {
		if current > 0 {
			return 100
		}
		return 0
	}
	return ((current - base) / base) * 100
}

func pctDeltaDown(base, current float64) float64 {
	if base <= 0 {
		return 0
	}
	if current >= base {
		return 0
	}
	return ((base - current) / base) * 100
}

func retainedPct(base, current float64) float64 {
	if base <= 0 {
		return 100
	}
	return (current / base) * 100
}

func bandByMax(actual, threshold float64) ValidationBand {
	if threshold <= 0 {
		return ValidationBandPass
	}
	if actual <= threshold {
		return ValidationBandPass
	}
	if actual <= threshold*1.2 {
		return ValidationBandWarn
	}
	return ValidationBandFail
}

func bandByMin(actual, threshold float64) ValidationBand {
	if threshold <= 0 {
		return ValidationBandPass
	}
	if actual >= threshold {
		return ValidationBandPass
	}
	if actual >= threshold*0.8 {
		return ValidationBandWarn
	}
	return ValidationBandFail
}

func maxBand(a, b ValidationBand) ValidationBand {
	order := map[ValidationBand]int{ValidationBandPass: 0, ValidationBandWarn: 1, ValidationBandFail: 2}
	if order[b] > order[a] {
		return b
	}
	return a
}

func minFloat(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}
